<a>/<area>'s origin getter returns "null" rather than "" when URL cannot be parsed

annevk opened the pull request:
#### 0e09f4e4a3b4803ec38c243cc8aaad0570cfd7fa
<pre>
&lt;a&gt;/&lt;area&gt;&apos;s origin getter returns &quot;null&quot; rather than &quot;&quot; when URL cannot be parsed
<a href="https://bugs.webkit.org/show_bug.cgi?id=128020">https://bugs.webkit.org/show_bug.cgi?id=128020</a>
rdar://114078288

Reviewed by Brent Fulgham.

Check if the URL cannot be parsed first (isValid() in our code) and if
so, return the empty string. This aligns us with the HTML Standard and
other browsers.

This is not implemented in URLDecomposition as URL, Location, and
WorkerLocation always represent parsable URLs.

* LayoutTests/fast/dom/HTMLAnchorElement/anchor-element-href-parsing-expected.txt:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-element-href-parsing.html:

Add coverage for username and password getters while here.

* Source/WebCore/html/HTMLAnchorElement.cpp:
(WebCore::HTMLAnchorElement::origin const):

Canonical link: <a href="https://commits.webkit.org/267222@main">https://commits.webkit.org/267222@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8465a8ec81cf6ac3d170b0ad1e2384c05ab98470

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15541 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15847 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16210 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17297 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14576 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15712 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/18364 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15942 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17137 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15726 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16166 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13216 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18041 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13434 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14018 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20950 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14482 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14184 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17456 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14772 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12515 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14028 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/13865 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3819 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18390 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14592 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/html/HTMLAnchorElement.cpp:
```diff
@@ -339,7 +339,10 @@ AtomString HTMLAnchorElement::target() const
-    return SecurityOrigin::create(href()).get().toString();
+    auto url = href();
+    if (!url.isValid())
+        return emptyString();
```
It's surprising there's no WPT for this. Should we propose one at push to wpt.fyi?

brentfulgham reviewed (approved):
Looks good. I wish there was a WPT impact for this.

annevk committed:
<a>/<area>'s origin getter returns "null" rather than "" when URL cannot be parsed
https://bugs.webkit.org/show_bug.cgi?id=128020
rdar://114078288

Reviewed by Brent Fulgham.

Check if the URL cannot be parsed first (isValid() in our code) and if
so, return the empty string. This aligns us with the HTML Standard and
other browsers.

This is not implemented in URLDecomposition as URL, Location, and
WorkerLocation always represent parsable URLs.

* LayoutTests/fast/dom/HTMLAnchorElement/anchor-element-href-parsing-expected.txt:
* LayoutTests/fast/dom/HTMLAnchorElement/anchor-element-href-parsing.html:

Add coverage for username and password getters while here.

* Source/WebCore/html/HTMLAnchorElement.cpp:
(WebCore::HTMLAnchorElement::origin const):

Canonical link: https://commits.webkit.org/267222@main
